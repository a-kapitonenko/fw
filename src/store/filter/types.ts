import { Frame } from '../frames/types';

export type Field = {
  name: string;
  value: string;
  checked?: boolean;
};

export type Groups = {
  color: Field[];
  width: Field[];
  noseBridge: Field[];
  shape: Field[];
  material: Field[];
};

export const enum FilterActionTypes {
  FETCH_REQUEST = '@@filter/FETCH_REQUEST',
  FETCH_SUCCESS = '@@filter/FETCH_SUCCESS',
  FETCH_ERROR = '@@filter/FETCH_ERROR',
  CHANGE_CHECKED = '@@filter/CHANGE_CHECKED',
};

export interface IFilterState {
  readonly fetching: boolean;
  readonly groups: Groups;
  readonly errors: string;
  readonly query: Object;
  readonly frames: Frame[];
};
