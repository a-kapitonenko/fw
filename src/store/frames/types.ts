export type Frame = {
  upc: number;
  label: string;
  value: string; 
  img: string;
  compatibility?: boolean;
};

export const enum FramesActionTypes {  
  SUBMIT_START = '@@frames/SUBMIT_START',
  SUBMIT_SUCCESS = '@@frames/SUBMIT_SUCCESS',
  SUBMIT_FAILED = '@@frames/SUBMIT_FAILED',
  FETCH_SIMILAR_FRAMES_START = '@@frames/FETCH_SIMILAR_FRAMES_START',
  FETCH_SIMILAR_FRAMES_SUCCESS = '@@frames/FETCH_SIMILAR_FRAMES_SUCCESS',
  FETCH_SIMILAR_FRAMES_FAILED = '@@frames/FETCH_SIMILAR_FRAMES_FAILED',
  CHECK_FRAME_START = '@@frames/CHECK_FRAME_START',
  CHECK_FRAME_SUCCESS = '@@frames/CHECK_FRAME_SUCCESS',
  CHECK_FRAME_FAILED = '@@frames/CHECK_FRAME_FAILED',
  OPEN = '@@frames/OPEN',
  CLOSE = '@@frames/CLOSE',
  SET_SELECTED_FRAME = '@@frames/SET_SELECTED_FRAME',
  CLEAR_SELECTED_FRAME = '@@frame/CLEAR_SELECTED_FRAME',
  CLEAR_RESULT = '@@frames/CLEAR_RESULT',
};

export interface FramesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly open: boolean;
  readonly selectedFrame: Frame;
  readonly similarFrames: {
    readonly isFetching: boolean;
    readonly errors: string;
    readonly data: Frame[];
  };
};
