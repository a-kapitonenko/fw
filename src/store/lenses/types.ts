export type Lens = {
  name: string;
  value: number;
};

export const enum LensesActionTypes {
  FETCH_LENSES_START = '@@lenses/FETCH_LENSES_START',
  FETCH_LENSES_SUCCESS = '@@lenses/FETCH_LENSES_SUCCESS',
  FETCH_LENSES_FAILED = '@@lenses/FETCH_LENSES_FAILED',

  SAVE_LENS_START = '@@lenses/SAVE_LENS_START',
  SAVE_LENS_SUCCESS = '@@lenses/SAVE_LENS_SUCCESS',
  SAVE_LENS_FAILED = '@@lenses/SAVE_LENS_FAILED',

  CHECK_LENS_START = '@@lenses/CHECK_LENS_START',
  CHECK_LENS_SUCCESS = '@@lenses/CHECK_LENS_SUCCESS',
  CHECK_LENS_FALIED = '@@lenses/CHECK_LENS_FAILED',
};

export interface ILensesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly lenses: Lens[];
};
