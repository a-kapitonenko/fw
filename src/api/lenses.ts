import request from './request';
import { Prescription, } from '../store/order/types';
import { Lens } from '../store/lenses/types';
import { saveLens, getLenses } from '../test/lenses';
import { checkLensError as checkLens } from '../test/lenses';
//import { checkLens } from '../test/lenses';

export const fetchLensesRequest = (data: Prescription) => {
  return request(getLenses, data);
};

export const saveLensRequest = (
  data: {
    prescription: Prescription,
    lens: Lens,
  }
) => {
  return request(saveLens, data);
};

export const checkLensRequest = (
  data: {
    prescription: Prescription,
    lens: Lens,
  }
) => {
  return request(checkLens, data);
};
