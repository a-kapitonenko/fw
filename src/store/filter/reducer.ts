import { Reducer } from 'redux';
import { IFilterState, FilterActionTypes, Groups, Field } from './types';

const initialGroupsState = {
  color: [],
  width: [],
  noseBridge: [],
  shape: [],
  material: []
}

const initialState: IFilterState = {
  isFetching: false,
  errors: '',
  query: initialGroupsState,
  frames: [],
  groups: {
    isFetching: false,
    errors: '',
    data: <Groups>initialGroupsState,
  }
};

const reducer: Reducer<IFilterState> = (state = initialState, action) => {
  switch (action.type) {
    case FilterActionTypes.FILTERING_START: {
      return { ...state, isFetching: true, errors: '' };
    }
    case FilterActionTypes.FILTERING_SUCCESS: {
      return { ...state, isFetching: false, frames: action.payload };
    }
    case FilterActionTypes.FILTERING_FAILED: {
      return { ...state, isFetching: false, errors: action.payload };
    }
    case FilterActionTypes.FETCH_GROUPS_START: {
      return { ...state, groups: { ...state.groups, isFetching: true, errors: '' } };
    }
    case FilterActionTypes.FETCH_GROUPS_SUCCESS: {
      return { ...state, groups: { ...state.groups, isFetching: false, data: action.payload } };
    }
    case FilterActionTypes.FETCH_GROUPS_FAILED: {
      return { ...state, groups: { ...state.groups, isFetching: true, errors: action.paylaod } };
    }
    case FilterActionTypes.CHANGE_CHECKED: {
      const item = state.groups.data[action.payload.type].map((field: Field) => {
        if (field.name === action.payload.name) {
          return { ...field, checked: action.payload.value };
        }

        return { ...field };
      });

      return {
        ...state,
        groups: {
          ...state.groups,
          data: {
            ...state.groups.data,
            [action.payload.type]: item
          }
        }
      };
    }
    case FilterActionTypes.CLEAR_CHECKED: {
      const groups = <Groups>{};

      for (const group in state.groups.data) {
        groups[group] = state.groups.data[group].map((field: Field) => ({ ...field, checked: false }));
      }

      return { ...state, groups: { ...state.groups, data: groups } };
    }
    case FilterActionTypes.ADD_QUERY: {
      return {
        ...state,
        query: {
          ...state.query,
          [action.payload.type]: [
            ...state.query[action.payload.type],
            action.payload.value
          ]
        }
      };
    }
    case FilterActionTypes.DELETE_QUERY: {
      const index = state.query[action.payload.type].indexOf(action.payload.value);

      return {
        ...state,
        query: {
          ...state.query,
          [action.payload.type]: [
            ...state.query[action.payload.type].slice(0, index),
            ...state.query[action.payload.type].slice(index + 1)
          ]
        }
      };
    }
    case FilterActionTypes.CLEAR_QUERY: {
      return { ...state, query: initialGroupsState };
    }
    default: {
      return state;
    }
  }
};

export { reducer as FilterReducer };
