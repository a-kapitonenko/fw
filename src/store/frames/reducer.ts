import { Reducer } from 'redux';
import { FramesState, FramesActionTypes, Frame } from './types';

const initialState: FramesState = {
  isFetching: false,
  errors: '',
  open: false,
  selectedFrame: <Frame>{},
  similarFrames: {
    isFetching: false,
    errors: '',
    data: []
  },
};

const reducer: Reducer<FramesState> = (state = initialState, action) => {
  switch (action.type) {
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_START: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: true, errors: '' } };
    }
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_SUCCESS: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: false, data: action.payload } };
    }
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_FAILED: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: false, errors: action.payload } };
    }
    case FramesActionTypes.SUBMIT_START: {
      return { ...state, isFetching: true, errors: '' };
    }
    case FramesActionTypes.SUBMIT_SUCCESS: {
      return { ...state, isFetching: false };
    }
    case FramesActionTypes.SUBMIT_FAILED: {
      return { ...state, isFetching: false, errors: action.payload };
    }
    case FramesActionTypes.OPEN: {
      return { ...state, open: true };
    }
    case FramesActionTypes.CLOSE: {
      return { ...state, open: false };
    }
    case FramesActionTypes.SET_SELECTED_FRAME: {
      return { ...state, selectedFrame: action.payload };
    }
    case FramesActionTypes.CLEAR_SELECTED_FRAME: {
      return { ...state, selectedFrame: <Frame>{} };
    }
    default: {
      return state;
    }
  }
};

export { reducer as FramesReducer };
