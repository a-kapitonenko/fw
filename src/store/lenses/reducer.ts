import { Reducer } from 'redux';
import { LensesState, LensesActionTypes } from './types';

const initialState: LensesState = {
  fetching: false,
  list: [],
  errors: '',
}

const reducer: Reducer<LensesState> = (state = initialState, action) => {
  switch (action.type) {
    case LensesActionTypes.FETCH_REQUEST: {
      return { ...state, fetching: true };
    }
    case LensesActionTypes.FETCH_SUCCESS: {
      return { 
        ...state, 
        fetching: false, 
        list: action.payload 
      };
    }
    case LensesActionTypes.FETCH_ERROR: {
      return { 
        ...state, 
        fetching: false, 
        errors: action.payload 
      };
    }
    default: {
      return state
    }
  }
};

export { reducer as LensesReducer }
