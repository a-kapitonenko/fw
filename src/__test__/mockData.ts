import { Frame } from '../store/frames/types';
import { Groups } from '../store/filter/types';
import { OculusInfo, Prescription } from '../store/order/types';
import { Lens } from '../store/lenses/types';

const oculusData: OculusInfo = {
  sphere: '1',
  cyclinder: '2',
  axis: '1',
  addition: '2',
  PDDistance: '1',
  PDNear: '1',
  prism: '1',
};

export const errorMessage: string = 'Error message';

export const prescription: Prescription = {
  OD: oculusData,
  OS: oculusData,
};

export const lenses: Lens[] = [
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown',
    value: 1
  },
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- Transitions* Signature VII',
    value: 2
  },
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- XTRActive* Gray',
    value: 3
  },
];

export const lens: Lens = {
  name: 'Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown',
  value: 1
};

export const frames: Frame[] = [
  {
    upc: 82850553738,
    label: 'FM17116',
    value: '82850553738',
    img: '1.png',
    compatibility: true,
  },
  {
    upc: 82850523999,
    label: 'SA17346',
    value: '82850523999',
    img: '2.png',
    compatibility: false,
  },
];

export const groups: Groups = {
  color: [
    { name: 'Black', value: 'black' },
  ],
  width: [
    { name: 'Narrow', value: 'narrow' },
  ],
  noseBridge: [
    { name: 'Standard', value: 'standard' },
  ],
  shape: [
    { name: 'Square', value: 'square' },
  ],
  material: [
    { name: 'Acetate', value: 'acetate' },
  ],
};

export const query: object = {
  color: ['black', 'gold', 'brown'],
  width: ['narrow', 'medium'],
  noseBridge: ['standard'],
  shape: ['square'],
  material: ['acetate'],
};
