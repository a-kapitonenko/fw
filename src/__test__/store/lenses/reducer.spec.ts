import { LensesActionTypes } from '../../../store/lenses/types';
import { LensesReducer } from '../../../store/lenses/reducer';
import * as mockData from '../../mockData';

describe('lenses reducer', () => {
  it('test handles FETCH_LENSES_START', () => {
    const filterAction = {
      type: LensesActionTypes.FETCH_LENSES_START,
      payload: {},
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles FETCH_LENSES_SUCCESS', () => {
    const filterAction = {
      type: LensesActionTypes.FETCH_LENSES_SUCCESS,
      payload: mockData.lenses,
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles FETCH_LENSES_FAILED', () => {
    const filterAction = {
      type: LensesActionTypes.FETCH_LENSES_FAILED,
      payload: mockData.errorMessage,
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles SAVE_LENS_START', () => {
    const filterAction = {
      type: LensesActionTypes.SAVE_LENS_START,
      payload: {},
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles SAVE_LENS_SUCCESS', () => {
    const filterAction = {
      type: LensesActionTypes.SAVE_LENS_SUCCESS,
      payload: {},
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles SAVE_LENS_FAILED', () => {
    const filterAction = {
      type: LensesActionTypes.SAVE_LENS_FAILED,
      payload: mockData.errorMessage,
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles CHECK_LENS_START', () => {
    const filterAction = {
      type: LensesActionTypes.CHECK_LENS_START,
      payload: {},
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles CHECK_LENS_SUCCESS', () => {
    const filterAction = {
      type: LensesActionTypes.CHECK_LENS_SUCCESS,
      payload: {},
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles CHECK_LENS_FALIED', () => {
    const filterAction = {
      type: LensesActionTypes.CHECK_LENS_FALIED,
      payload: mockData.errorMessage,
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });

  it('test handles SET_INITIAL_VALUE', () => {
    const filterAction = {
      type: LensesActionTypes.SET_INITIAL_VALUE,
      payload: mockData.lenses,
    };
    
    expect(LensesReducer(undefined, filterAction)).toMatchSnapshot();
  });
});
