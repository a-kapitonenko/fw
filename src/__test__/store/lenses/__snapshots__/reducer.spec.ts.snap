// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lenses reducer test handles CHECK_LENS_FALIED 1`] = `
Object {
  "errors": "Error message",
  "isFetching": false,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles CHECK_LENS_START 1`] = `
Object {
  "errors": "",
  "isFetching": true,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles CHECK_LENS_SUCCESS 1`] = `
Object {
  "errors": "",
  "isFetching": false,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles FETCH_LENSES_FAILED 1`] = `
Object {
  "errors": "Error message",
  "isFetching": false,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles FETCH_LENSES_START 1`] = `
Object {
  "errors": "",
  "isFetching": true,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles FETCH_LENSES_SUCCESS 1`] = `
Object {
  "errors": "",
  "isFetching": false,
  "lenses": Array [
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown",
      "value": 1,
    },
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- Transitions* Signature VII",
      "value": 2,
    },
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- XTRActive* Gray",
      "value": 3,
    },
  ],
}
`;

exports[`lenses reducer test handles SAVE_LENS_FAILED 1`] = `
Object {
  "errors": "Error message",
  "isFetching": false,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles SAVE_LENS_START 1`] = `
Object {
  "errors": "",
  "isFetching": true,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles SAVE_LENS_SUCCESS 1`] = `
Object {
  "errors": "",
  "isFetching": false,
  "lenses": Array [],
}
`;

exports[`lenses reducer test handles SET_INITIAL_VALUE 1`] = `
Object {
  "errors": "",
  "isFetching": false,
  "lenses": Array [
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown",
      "value": 1,
    },
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- Transitions* Signature VII",
      "value": 2,
    },
    Object {
      "name": "Nikon Eyes* Advanced + Polycarbonate- XTRActive* Gray",
      "value": 3,
    },
  ],
}
`;
